### PR TITLE
Case SQL functions

### PR DESCRIPTION
**Pull request by snopoke**
@NoahCarnahan :blowfish:

### CODE REVIEW

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 1]
```diff
@@ -153,82 +151,85 @@ def case_modified_since(case_id, server_modified_on):
...
     @staticmethod
     def get_case_by_location(domain, location_id):
         try:
-            return CommCareCaseSQL.objects.filter(
-                domain=domain,
-                type='supply-point',
-                location_id=location_id
-            ).get()
-        except CommCareCaseSQL.DoesNotExist:
+            return CommCareCaseSQL.objects.raw('SELECT * from get_case_by_location_id(%s, %s)', [domain, location_id])[0]
```
Comment (dimagimon) [inline corehq/form_processor/backends/sql/dbaccessors.py]: E501 line too long (121 > 115 characters)

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 2]
```diff
@@ -239,3 +240,11 @@ def _order_list(id_list, object_list, id_property):
...
 
     return ordered_list
+
+
+def _attach_prefetch_models(objects_by_id, prefetched_models, link_field_name, cached_attrib_name):
+    prefetched_groups = groupby(prefetched_models, lambda x: getattr(x, link_field_name))
+    for obj_id, group in prefetched_groups:
+        obj = objects_by_id[obj_id]
+        setattr(obj, cached_attrib_name, list(group))
+
```
Comment (dimagimon) [inline corehq/form_processor/backends/sql/dbaccessors.py]: W391 blank line at end of file

File: corehq/form_processor/tests/test_case_dbaccessor.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    form = XFormInstanceSQL(
+        form_id=form_id,
+        xmlns='http://openrosa.org/formdesigner/form-processor',
+        received_on=utcnow,
+        user_id=user_id,
+        domain=domain
+    )
+
+    cases = []
+    if case_id:
```
Comment (NoahCarnahan) [inline corehq/form_processor/tests/test_case_dbaccessor.py]: Will this ever be false?

File: corehq/form_processor/tests/test_case_dbaccessor.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+        case_ids = CaseAccessorSQL.get_case_ids_in_domain(DOMAIN)
+        self.assertEqual({case1.case_id, case3.case_id}, set(case_ids))
+
+
+def _create_case(domain=None, form_id=None, case_type=None):
+    """
+    Create the models directly so that these tests aren't dependent on any
+    other apps. Not testing form processing here anyway.
+    :return: case_id
```
Comment (NoahCarnahan) [inline corehq/form_processor/tests/test_case_dbaccessor.py]: Very minor, but looks like this returns a case, not an id.

Comment (NoahCarnahan) [inline corehq/form_processor/tests/test_case_dbaccessor.py]: > so that these tests aren't dependent on any other apps

woohoo!!

Comment (czue) [inline corehq/form_processor/tests/test_case_dbaccessor.py]: wow that's brilliant. I should do that more for stuff that doesn't actually test form processing.

File: corehq/form_processor/tests/test_case_dbaccessor.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        self.assertEqual(2, len(cases))
+        self.assertEqual(case1.case_id, cases[0].case_id)
+        self.assertEqual(case2.case_id, cases[1].case_id)
+
+    def test_case_modified_since(self):
+        case = _create_case()
+
+        self.assertFalse(CaseAccessorSQL.case_modified_since(case.case_id, case.server_modified_on))
+
+        self.assertTrue(CaseAccessorSQL.case_modified_since(case.case_id, datetime.utcnow()))
```
Comment (NoahCarnahan) [inline corehq/form_processor/tests/test_case_dbaccessor.py]: Should this actually return false? To me this says: "Has the case been modified after `datetime.utcnow()`?", and I think the answer to that question would be "no". I might be misunderstanding the meaning of `case_modified_since()` or something else though...

File: corehq/form_processor/sql_functions/get_case_form_ids.sql [hunk 6]
```diff
@@ -0,0 +1,16 @@
+DROP FUNCTION IF EXISTS get_case_form_ids(text);
+
+CREATE FUNCTION get_case_form_ids(case_id text) RETURNS TABLE (form_id VARCHAR(255)) AS $$
```
Comment (NoahCarnahan) [inline corehq/form_processor/sql_functions/get_case_form_ids.sql]: Would it be more appropriate for form_id type to be `text`?

Comment (NoahCarnahan) [inline corehq/form_processor/sql_functions/get_case_form_ids.sql]: I don't not a whole lot about pgsql, but I believe it's possible to use `SETOF text` or something like that for a function return type, which I think would achieve the same thing as returning the table here. Do you have any idea if one approach is more performant than the other? My intuition is that returning a table is a more heavy duty operation...

**Comment by NoahCarnahan** [comment]
Looks good. No blockers from me, but added a few questions. Another question: Are you planning on overriding the object Manager to disallow functions other than `raw()`? Seems like it would be a good idea to ensure that other methods aren't being used now or in the future.

**Pull request merged**